src/System.Private.CoreLib/shared/System/Threading/ThreadLocal.cs

```diff
@@ -356,7 +358,8 @@ private void SetValueSlow(T value, LinkedSlotVolatile[] slotArray)
             // If the slot array is not big enough to hold this ID, increase the table size.
             if (id >= slotArray.Length)
             {
-                GrowTable(ref slotArray, id + 1);
+                GrowTable(ref slotArray!, id + 1);
+                Debug.Assert(ts_finalizationHelper != null, "Should have been initialized when this thread's slot array was created.");
```

Comment:
I haven't figured out how I feel about all these new asserts. I don't think we consistently assert in such cases outside of this annotation work. Not sure whether it makes the code more readable, or less. It does seem the best way to resolve nullability flow though.

Reply:
It would be nice if the compiler would flag when a `!` marker was _not_ necessary (which would happen if it became unnecessary as a side effect of subsequent refactoring). If that was the case, at that point it might be an advantage of using `!` over `Debug.Assert`.